### DISCUSSION TITLE
World loading with Fabric

### DISCUSSION THREAD

**Discussion by runi95** [runi95/turtle-control-panel#12] [Ideas]
Discussion on how to get proper world loading, exploration and discovery working for those who use the Fabric launcher.

**Comment by SolsticeSpectrum** [comment]
I am getting a lot of these
![image](https://github.com/runi95/turtle-control-panel/assets/38611351/f95e7e4f-f721-44d7-9b39-b8b0aed7b55c)

Initially I tried deleting the assets with these but now I stumbled on one where `promenade` isn't existing asset folder
```
Skipping: elite_coffin_open.json - missing elements
Skipping: lily_pad.json - missing textures
Failed to load textures for model: blush_sakura_small_hedge_end
/home/ruda0/Github/turtle-control-panel/atlas-generator/node_modules/canvas/lib/image.js:94
  SetSource.call(img, src)
```
I think it should skip it but it doesn't work with some

EDIT: I added this which worked nicely
```js
    if (!fs.existsSync(filePath)) {
        console.log(`Skipping: ${filePath} - missing path`);
        continue;
    }
 ```

**Comment by SolsticeSpectrum** [comment]
Also slabs are not placed correctly and leaves are gray probably due to ignoring of `distance` state
And the sprite generation would be welcome

**Comment by runi95** [comment]
@DarkReaper231 Not quite sure what you mean by "slabs are not placed correctly", but I've experienced some placement issues with blocks before. All blocks that can be rotated (furnace, stairs, saw, etc) will always only have 1 rotation, because when a turtle inspects a block we only get the position (x, y, z) and the block name, nothing about block rotation. This means all block rotations will always be wrong.

Another issue we have due to how turtles see the world is stuff that can be 1/2 block up or down, the blocks I'm aware of that can do this are snow blocks and slabs. The problem is that the turtle will only give us the x, y, z coordinates, but it never gives half coordinates. So we will never know if a slab is supposed to touch the ground or be 1/2 block up.

As for the texture loading issues you have we might just have to skip them like you're doing now or it might be a potential improvement we can make within the atlas-generator.
So far I've only ever used it to generate textures for Minecraft, CC:Tweaked and Advanced Peripherals meaning there might be a whole bunch of stuff from other mods that does not load correctly. I'll see if I can try loading the promenade mod sometime and see what might be the issue for the atlas-generator or the textures might simply be missing and impossible to load fully load.

**Comment by SolsticeSpectrum** [comment]
@runi95 The promenade mod is not even in my mods folder. I think some other mod is just trying to use it if it's available. For some mods there are paths that just don't exist, I assume it's because they're dynamic and get created in memory instead of being physically present in assets folder. One example I found like this is the chizel mod. Skipping these missing paths is the best options because we don't care about these edge cases.

**Reply by runi95** [reply]
Alright, good to know! 😄

**Comment by SolsticeSpectrum** [comment]
@runi95 Okay here is the result of my efforts, is it possible to do something about the missing grass and leave colors?
![image](https://github.com/runi95/turtle-control-panel/assets/38611351/6d7b7022-f9b9-4eb7-8199-292c8bd38ce4)

**Reply by runi95** [reply]
Oh-yeah, I forgot to answer why the leaves and stuff is off-color. So the reason it's like this is because some blocks change color depending on the biome they're in. What Minecraft does here is that they create a colorless texture (what we've rendered here) and then simply add color to them manually afterwards.
The colors are nowhere in the assets folder, but I think it might be hardcoded into Minecraft somewhere? Not too sure about exactly where it's at actually. It's definitely applied to the model afterwards though as the color is unknown until the block is placed on the ground. Only then do they know the biome and thus the color to use. The asset folder does say something about `tint` which lets us know the asset is missing color, but doesn't tell us which could it should be.

**Reply by runi95** [reply]
I've manually fixed this for some of the blocks by manually adding colors to the textures in the assets directory for them. This isn't going to work when you dynamically load and create the textures like this and a better solution would be to figure out how to find the correct color anyway. I just don't know where this information sits.
Making everything that's missing color green would be a fair approximation as green tends to be the standard color for most of these blocks as far as I know at least.

**Reply by runi95** [reply]
Oh-wait, no green won't actually work as both water and lava have dynamic coloring depending on biome. They're also just grey in their texture files... 😅

**Reply by SolsticeSpectrum** [reply]
What about checking if block is water or grass and setting color. Won't probably work for some crazy colorful modded biomes but that ain't that big of a deal

**Reply by SolsticeSpectrum** [reply]
Also what's happening with the water? It has missing texture.

**Reply by runi95** [reply]
Yeah, both water and lava are pretty weird as their model files don't actually include any model definitions. I think the reason for this is because they're liquids and don't really have a proper model structure, they probably have some custom rendering logic somewhere within Minecraft instead. In addition to this their texture files are colorless.

So what I did is I simply created a model for them based off the cube definition. This works for source blocks, but no other types of water/lava. Since turtles can only see source blocks this works pretty well anyway.

Another rendering issue I've had is that entities like carts, boats and chests don't actually have model definitions in assets and are also handled somewhere within Minecraft's source code.
Usually entities are stuff like mobs and turtles can't see those as blocks anyway so it works fine... except that chests are entities **and** blocks at the same time and thus can be seen by the turtles yet not rendered without some help.

What I've done for all these exceptions so far has been creating custom models for these, but since this "custom" model is just the Minecraft block model with the original water, lava and chest textures I can't actually upload those custom models to GitHub.

**Reply by runi95** [reply]
I think I have an idea for how to fix this, but the rendering for chests is pretty difficult due to the textures width not being a multiplier of 16, but rather 14. This messes up ThreeJS's built-in atlas which is what we're using to render everything.

**Reply by runi95** [reply]
@DarkReaper231 Okay, I've implemented the suggested logic. I've created a custom model override for water and lava as they have no model definitions by default in the Minecraft assets directory. I've also added a green color to all blocks that have a `tintindex` defined in their texture definition.

This solution should fix pretty much all the color issues you've had as well as the rendering of water and lava, **but** it broke the rendering of chests which we're going to have to fix and handle separately in the rendering logic somewhere later down the line.
